module/VuFind/src/VuFind/Db/Service/UserResourceService.php

```diff
@@ -86,4 +89,54 @@ public function getStatistics(): array
     {
         return $this->getDbTable('UserResource')->getStatistics();
     }
+
+    /**
+     * Create user/resource/list link if one does not exist; update notes if one does.
+     *
+     * @param ResourceEntityInterface|int $resourceOrId Entity or ID of resource to link up
+     * @param UserEntityInterface|int     $userOrId     Entity or ID of user creating link
+     * @param UserListEntityInterface|int $listOrId     Entity or ID of list to link up
+     * @param string                      $notes        Notes to associate with link
+     *
+     * @return UserResource|false
+     */
+    public function createOrUpdateLink(
+        ResourceEntityInterface|int $resourceOrId,
+        UserEntityInterface|int $userOrId,
+        UserListEntityInterface|int $listOrId,
+        string $notes = ''
+    ): UserResourceEntityInterface {
+        $resource = $resourceOrId instanceof ResourceEntityInterface
+            ? $resourceOrId : $this->getDbService(ResourceServiceInterface::class)->getResourceById($resourceOrId);
+        $list = $listOrId instanceof UserListEntityInterface
+            ? $listOrId : $this->getDbService(UserListServiceInterface::class)->getUserListById($listOrId);
+        $user = $userOrId instanceof UserEntityInterface
+            ? $userOrId : $this->getDbService(UserServiceInterface::class)->getUserById($userOrId);
```

Comment:
Need to check that all were found and no nulls were returned, right? Otherwise if an entity isn't in the database, the code below will blow.

Reply:
None of these parameters are nullable at present. The current code doesn't require them to be, though I could add more flexibility if you think it might be necessary in future.

Reply:
No, but e.g. `getResourceById` coudl return null.

Reply:
Ahh, sorry, I misunderstood. Good point; I think we'll have to throw an exception if that occurs, but that's better than unexpectedly blowing up. :-)

Reply:
Okay, addressed by 489a23e2471000fc02d5fe1e1fa3e489ba99697b.